Modules/blocks_output_buffer.h

```diff
@@ -0,0 +1,332 @@
+/*
+   _BlocksOutputBuffer is used to maintain an output buffer
+   that has unpredictable size. Suitable for compress/decompress
+   API (bz2/lzma/zlib) that has stream->next_out and stream->avail_out:
+
+        stream->next_out:  point to the next output position.
+        stream->avail_out: the number of available bytes left in the buffer.
+
+   It maintains a list of bytes object, so there is no overhead
+   of resizing the buffer.
+
+   Usage:
+
+   1, Define BOB_BUFFER_TYPE and BOB_SIZE_TYPE, and include this file:
+        #define BOB_BUFFER_TYPE uint8_t  // type of next_out pointer
+        #define BOB_SIZE_TYPE uint32_t   // type of avail_out
+        #include "blocks_output_buffer.h"
+
+   2, Initialize the buffer use one of these functions:
+        _BlocksOutputBuffer_Init()
+        _BlocksOutputBuffer_InitAndGrow()
+        _BlocksOutputBuffer_InitWithSize()
+
+   3, If (avail_out == 0), grow the buffer:
+        _BlocksOutputBuffer_Grow()
+
+   4, Get the current outputted data size:
+        _BlocksOutputBuffer_GetDataSize()
+
+   5, Finish the buffer, and return a bytes object:
+        _BlocksOutputBuffer_Finish()
+
+   6, Clean up the buffer when an error occurred:
+        _BlocksOutputBuffer_OnError()
+*/
+
+
+typedef struct {
+    /* List of blocks */
+    PyObject *list;
+    /* Number of whole allocated size. */
+    Py_ssize_t allocated;
+
+    /* Max length of the buffer, negative number for unlimited length. */
+    Py_ssize_t max_length;
+} _BlocksOutputBuffer;
+
+#if defined(__GNUC__)
+#  define UNUSED_ATTR __attribute__((unused))
+#else
+#  define UNUSED_ATTR
+#endif
+
+#if BOB_SIZE_TYPE == uint32_t
+#  define BOB_SIZE_MAX UINT32_MAX
+#elif BOB_SIZE_TYPE == size_t
+#  define BOB_SIZE_MAX UINTPTR_MAX
+#endif
+
+/* Block size sequence. Some compressor/decompressor can't process large
+   buffer (>4GB), so the type is int. Below functions assume the type is int.
+*/
+#define KB (1024)
+#define MB (1024*1024)
+static const int BUFFER_BLOCK_SIZE[] =
+    { 32*KB, 64*KB, 256*KB, 1*MB, 4*MB, 8*MB, 16*MB, 16*MB,
+      32*MB, 32*MB, 32*MB, 32*MB, 64*MB, 64*MB, 128*MB, 128*MB,
+      256*MB };
+#undef KB
+#undef MB
+
+/* According to the block sizes defined by BUFFER_BLOCK_SIZE, the whole
+   allocated size growth step is:
+    1   32 KB       +32 KB
+    2   96 KB       +64 KB
+    3   352 KB      +256 KB
+    4   1.34 MB     +1 MB
+    5   5.34 MB     +4 MB
+    6   13.34 MB    +8 MB
+    7   29.34 MB    +16 MB
+    8   45.34 MB    +16 MB
+    9   77.34 MB    +32 MB
+    10  109.34 MB   +32 MB
+    11  141.34 MB   +32 MB
+    12  173.34 MB   +32 MB
+    13  237.34 MB   +64 MB
+    14  301.34 MB   +64 MB
+    15  429.34 MB   +128 MB
+    16  557.34 MB   +128 MB
+    17  813.34 MB   +256 MB
+    18  1069.34 MB  +256 MB
+    19  1325.34 MB  +256 MB
+    20  1581.34 MB  +256 MB
+    21  1837.34 MB  +256 MB
+    22  2093.34 MB  +256 MB
+    ...
+*/
```

Comment:
Why were these specific sizes chosen?

Reply:
This code seems to have linear performance, so these sizes should not affect speed, only affect memory usage.

I tried to adjust them to suit compression/decompression scenarios.

For example, there is overhead when frequently calling the underlying library's compress/decompress function, so the growth step is not very small.

The initial size is 32KB. Some code in (or not in) stdlib use 8KB read buffer, if the compression ratio <= 4 [1], then it can be decompressed within one .decompress() call.

If other reasons are found, they can be adjusted.

[1] https://cran.r-project.org/web/packages/brotli/vignettes/brotli-2015-09-22.pdf